Interpret `type` annotations as `type[Any]` annotations

tyralla opened the pull request:
Fixes #16349

I had to adjust three existing tests.  In my opinion, all necessary adjustments indicate an increase in consistency.

tyralla commented:
Regarding the Mypy primer's output, most changes inform about missing type arguments for `type` (as expected).  I will check the other changes later.

All the failing Mypyc test suits also seem to be due to missing type arguments.  No idea how many such problems exist.  I am not familiar with Mypyc, but I suppose that inserting the most precise type possible leads to higher efficiency than just inserting `Any`?

tyralla commented:
I had a look at the Mypy primer diff.  Most differences are, as expected, complaints about missing type arguments and error messages that now contain `type[Any]` instead of `type`, so that's okay.   The other differences are mainly due to Mypy now understanding that `object` is no subtype of `type`, which also looks like a relevant improvement. 

So, I think this pull request is ready for review.

JelleZijlstra reviewed (changes requested):
This is an improvement but unfortunately I think it's not quite right yet. The trouble is that `type` sometimes *should* be interpreted as `type[Any]` even if it's not syntactically within an annotation.

For example, consider this:

```
from typing import Type
class L1(list[type]): ...
class L2(list[Type]): ...

l1: L1
for elt in l1:
    reveal_type(elt)
l2: L2
for elt in l2:
    reveal_type(elt)
```

With your PR, when I put this in a test case I get
```
  main:36: note: Revealed type is "builtins.type" (diff)
  main:39: note: Revealed type is "Type[Any]" (diff)
```
While I would expect both to behave the same way.

I haven't tried it but I suspect you'll see similar issues in other places where types can appear outside annotations (e.g., `cast()`, `TypeVar` bounds).

Intuitively, I think the right behavior is that `analyze_annotation` should always be true. However, when I tried that, I got a ton of test failures about inheriting from `type`. Maybe there's a way to fix that, though.

tyralla commented:
You are right; addressing all places where `type` can be typed would be a bigger improvement.  I will give it a try...

tyralla commented:
After some thinking and reading, I think `analyze_annotation` should be more often but not always true.  It should be only true if `type` could be replaced with `Type` (for now).  In my understanding, `type` (unfortunately) currently has three different meanings:

First, it allows us to define that we want to handle the class of an instance and not the instance itself.  That is what this pull request is about, and I clearly should support Jelle's example, `cast`, and so on.

Second, it serves as a function to query the type of an object.  In this regard, `type(instance)` is somehow like `type[object](instance)` and also like `type[Any](instance)` because it accepts instances of any type, and every type is a subtype of object.  But forcing users to write `type[object](instance)` seems weird.  (Then, what would `type[int](instance)` mean?  That `instance` must be a subtype of `int`, and type checkers should otherwise throw an error?)  So here, I guess `analyze_annotation` should be false for now.

Third, it serves as a factory function for generating class objects.  Here, anything I can think of is that `T` in `type[T]` could mean the return type.  The following code works, but Mypy complains:
 
```python
class A: x = 1
B = type("B", (A,), {})
print(B.x)  # error: "type" has no attribute "x"
```

So, here it could eventually be helpful to annotate with `A`, but then Mypy says:

```python
class A: x = 1
B = type[A]("B", (A,), {})  # error: Value of type "type[type]" is not indexable
```

I am sure others have thought this through much more thoroughly, and this is way beyond my current intention.  Hence, for this pull request, `analyze_annotation` should also be false in such cases for now, in my opinion.

JelleZijlstra commented:
Some quick notes (will have a full review later):

* The `jax` hit in mypy-primer above is concerning, since it seems to indicate a bug where mypy thinks types are not hashable. That needs to be fixed.
* I'm not convinced by your argument for why `analyze_annotation` shouldn't always be true, because in the contexts you outline, we shouldn't be calling `anal_type` in the first place (since these contexts do not hold types).
* I don't like the name `analyze_annotation`, I think it should be something like `is_annotation`, or even `is_type` since we also use it outside of annotations.

tyralla commented:
> Some quick notes (will have a full review later):

Thanks!

> * The `jax` hit in mypy-primer above is concerning, since it seems to indicate a bug where mypy thinks types are not hashable. That needs to be fixed.

Yes, and there are some more problems I have to think about, too.  I will do it later.

> * I'm not convinced by your argument for why `analyze_annotation` shouldn't always be true, because in the contexts you outline, we shouldn't be calling `anal_type` in the first place (since these contexts do not hold types).

I misunderstood and thought you suggested to *always* convert `type` to `type[Any]`.  I will rethink this.

> * I don't like the name `analyze_annotation`, I think it should be something like `is_annotation`, or even `is_type` since we also use it outside of annotations.

Okay, let's decide on the name as soon as everything works as desired.

tyralla commented:
I added a commit, but the test machinery does not start.

Anyhow, I  renamed `analyze_annotation` to `builtin_type_is_type_type` (long but relatively explicit, IMO) and made it true by default.   There are only very few exceptions related to:

```python
t = type
class A(metaclass=t)

class B(type): ...
```

This new behaviour is debatable (conflicts with dataclasses.pyi):

```python
_InitVarMeta: TypeAlias = type  # E: Missing type parameters for generic type "type"
class InitVar(Generic[_T], metaclass=_InitVarMeta):  # E: Invalid metaclass "_InitVarMeta"
```

Also, I added a new error message that tries to clarify things:

```python
type[str](1)  # E: "builtins.type" is indexable as a type hint but neither a generic class nor a generic function
r = type[str]("r", (), {})  # E: "builtins.type" is indexable as a type hint but neither a generic class nor a generic function
class S(type[str]): ...  # E: "builtins.type" is indexable as a type hint but neither a generic class nor a generic function
```

I would look after the hashability problem and check if the code changes could be further simplified after we agreed on (or modified) these functionalities.

JelleZijlstra commented on test-data/unit/check-generics.test:
```diff
@@ -513,7 +513,7 @@ Alias[int]("a")  # E: Argument 1 to "Node" has incompatible type "str"; expected
 
 [case testTypeApplicationCrash]
-type[int] # this was crashing, see #2302 (comment)  # E: The type "Type[type]" is not generic and not indexable
+type[int] # this was crashing, see #2302 (comment)  # E: "builtins.type" is indexable as a type hint but neither a generic class nor a generic function
```
I wouldn't expect this to error at all on 3.9+.

tyralla replied:
I made this message for the following "errors" (taken from the `testBuiltinTypeType` test case):

```python
type[str](1)  # E: "builtins.type" is indexable as a type hint but neither a generic class nor a generic function
r = type[str]("r", (), {})  # E: "builtins.type" is indexable as a type hint but neither a generic class nor a generic function
class S(type[str]): ...  # E: "builtins.type" is indexable as a type hint but neither a generic class nor a generic function
```

JelleZijlstra commented on test-data/unit/check-newsemanal.test:
```diff
@@ -3215,6 +3215,25 @@ class User:
         self.name = name  # E: Cannot assign to a method
 
+
+
```
Too many blank lines

tyralla replied:
Committed accidentally; I will remove them.

JelleZijlstra commented on test-data/unit/check-newsemanal.test:
```diff
@@ -3241,3 +3260,52 @@ class b:
+
+O = TypeVar("O", bound=type)  # E: Missing type parameters for generic type "type"
+def p(q: O) -> O: ...
+reveal_type(str)  # N: Revealed type is "def () -> builtins.str"
```
Did you mean `reveal_type(p(str))`?

tyralla replied:
Yes, of course, thanks!

JelleZijlstra commented on test-data/unit/fixtures/dict.pyi:
```diff
@@ -77,7 +87,7 @@ class bool(int): pass
 class ellipsis:
     __class__: object
-def isinstance(x: object, t: Union[type, Tuple[type, ...]]) -> bool: pass
+def isinstance(x: object, t: Union[type[Any], Tuple[type[Any], ...]]) -> bool: pass  # type: ignore
```
Why does this need a type ignore? I thought it might be that mypy didn't think `type` was indexable, but it seems fine in the tuple fixture below.

tyralla replied:
`testErrorCodeMissingMultiple` demands error codes in type ignores.  I added them.

JelleZijlstra reviewed (commented):
Thanks, this is looking like a step in the right direction.

tyralla commented:
> > This new behaviour is debatable (conflicts with dataclasses.pyi):
> 
> The new behavior feels right. The metaclass is a runtime concept; we don't need to allow using a type alias there.

Yes, I thought so, too.  Okay, then I will leave it like this and will next have a look at the hashability problem.

tyralla commented:
> * The `jax` hit in mypy-primer above is concerning, since it seems to indicate a bug where mypy thinks types are not hashable. That needs to be fixed.

I checked if this is an already existing limitation.  Mypy 1.7 reports:

```
from typing import Any, Hashable
t: type[Any]
h: Hashable = t  # E: Incompatible types in assignment (expression has type "type[Any]", variable has type "Hashable")  [assignment]
```

So, should we tackle this issue in a separate PR?

tyralla commented:
Hashing, protocols and metaclasses - not a nice combination...  

After much debugging, I tried a very direct approach that at least considers that someone could overwrite `__hash__` with `None` in a metaclass.

JelleZijlstra commented on mypy/subtypes.py:
```diff
@@ -1055,6 +1055,20 @@ def visit_type_type(self, left: TypeType) -> bool:
+            if (
+                right.type.is_protocol
+                and (len(right.type.protocol_members) == 1)
+                and (right.type.protocol_members[0] == "__hash__")
```
This would behave incorrectly in an exotic case where the protocol contains a member named `__hash__` that doesn't have the standard signature. Maybe that's OK though as it's better than the current state.

tyralla replied:
The same holds for the potential subclass.  I added additional checks for both of them.

JelleZijlstra commented on test-data/unit/check-newsemanal.test:
```diff
@@ -3241,3 +3241,52 @@ class b:
+
+O = TypeVar("O", bound=type)  # E: Missing type parameters for generic type "type"
+def p(q: O) -> O: ...
+reveal_type(p(str))  # N: Revealed type is "def () -> builtins.str"
```
Add a test that a non-type is rejected, e.g. `p(1)  # E: <something about incompatible types>`

tyralla replied:
Okay, I add `p(1)  # E: Value of type variable "O" of "p" cannot be "int"`.

JelleZijlstra commented on mypy/subtypes.py:
```diff
@@ -1055,6 +1055,30 @@ def visit_type_type(self, left: TypeType) -> bool:
+                    return True
+                if isinstance(item, Instance):
+                    if (mtype := item.type.metaclass_type) is None or (
+                        symtab := mtype.type.get("__hash__")
```
symtab looks unused here

tyralla replied:
I removed it.

JelleZijlstra reviewed (commented):
This is looking good, but the amount of primer hits is concerning (cc @JukkaL). I think this is the right thing to do, but it's prudent to wait until the Typing Council decides on `type/Type` semantics (https://discuss.python.org/t/inconsistencies-between-type-and-type/37404/12), and we may need some mechanism to make migration easier, maybe an error code specifically for bare `type`.

tyralla commented:
@JelleZijlstra @JukkaL It seems the Typing Council decided in agreement with this PR; at least the [typing specs](https://typing.readthedocs.io/en/latest/spec/special-types.html#type) indicate so:

> Plain type without brackets, the root of Python's metaclass hierarchy, is equivalent to type[Any].

So, a possible migration mechanism is the only issue left for this PR, isn't it?  Personally, I think that adding the missing parameters (or just `Any`)  should not be too much effort, even for large code bases.  However, I could try to add a specific error code which allows silencing the `Missing type parameters for generic type "type"` errors and inserts `Any` automatically.
